plugins/radar-android-google-sleep/src/main/java/org.radarbase.passive.google.sleep/GoogleSleepManager.kt

```diff
@@ -0,0 +1,153 @@
+
+    private val sleepBroadcastReceiver: BroadcastReceiver
+    private val sleepHandler = SafeHandler.getInstance("Google Sleep", Process.THREAD_PRIORITY_BACKGROUND)
+    private lateinit var sleepPendingIntent: PendingIntent
+
+    init {
+        name = context.getString(R.string.googleSleepDisplayName)
+        sleepBroadcastReceiver = SleepReceiver(this)
+    }
+
+    override fun start(acceptableIds: Set<String>) {
+        register()
+        sleepHandler.start()
+        status = SourceStatusListener.Status.READY
+        sleepHandler.execute {
+            createSleepPendingIntent()
+            registerForSleepData()
+        }
+    }
+
+     fun sendSleepClassifyData(sleepIntent: Intent) {
+        logger.info("Sleep segment event data received")
+        val sleepSegmentEvents: List<SleepSegmentEvent> = SleepSegmentEvent.extractEvents(sleepIntent)
+        sleepSegmentEvents.forEach { sleepSegmentEvent ->
+            val time = currentTime
+            val sleepStartTime: Double = sleepSegmentEvent.startTimeMillis / 1000.0
+            val sleepEndTime: Double = sleepSegmentEvent.endTimeMillis / 1000.0
+            val sleepDuration: Double = sleepSegmentEvent.segmentDurationMillis / 1000.0
+            val sleepStatus: SleepStatus = sleepSegmentEvent.status.toSleepStatus()
+
+            send(segmentEventTopic, GoogleSleepSegmentEvent(time, time, sleepStartTime,
+            sleepEndTime, sleepDuration, sleepStatus))
+        }
+    }
+
+     fun sendSleepSegmentData(sleepIntent: Intent) {
+        logger.info("Sleep classify event data received")
+        val sleepClassifyEvents: List<SleepClassifyEvent> = SleepClassifyEvent.extractEvents(sleepIntent)
+        sleepClassifyEvents.forEach {  sleepClassifyEvent ->
+            val time = sleepClassifyEvent.timestampMillis / 1000.0
+            val sleepConfidence = sleepClassifyEvent.confidence
+            val light = sleepClassifyEvent.light
+            val motion = sleepClassifyEvent.motion
+
+            send(classifyEventTopic, GoogleSleepClassifyEvent(time, currentTime, sleepConfidence, light, motion))
+        }
+    }
+
+    @SuppressLint("MissingPermission")
+    private fun registerForSleepData() {
+        if ( isPermissionGranted() ) {
+            status = SourceStatusListener.Status.CONNECTING
+
+            ActivityRecognition.getClient(service)
+                .requestSleepSegmentUpdates(
+                    sleepPendingIntent,
+                    SleepSegmentRequest.getDefaultSleepSegmentRequest() )
+                .addOnSuccessListener {
+                    status = SourceStatusListener.Status.CONNECTED
+                    logger.info("Successfully subscribed to sleep data")
+                }
+                .addOnFailureListener {exception ->
+                    status = SourceStatusListener.Status.UNAVAILABLE
```

Comment:
This will cause the plugin to be permanently unavailable after running once. Should you check if this is a temporary failure? If so, use `disconnect()` instead.